### PR TITLE
Feature/add codespell

### PR DESCRIPTION
**Pull request by cryptomental**
This pull request solves #4442 

* Integrate codespell to Circle CI
* Fix typos
* Create codespell whitelist with known words that should not be considered as misspelled.

### CODE REVIEW

**Comment by cryptomental** [comment]
Documentation failed on a malformed table -> fix in progress.

**Comment by cryptomental** [comment]
I will double check the failing tests.

File: scripts/codespell_whitelist.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+iff
+nd
+assignend
```
Comment (axic) [inline scripts/codespell_whitelist.txt:3]: Is this (and `nd`) a real world?

File: .circleci/config.yml [hunk 2]
```diff
@@ -207,11 +211,14 @@ jobs:
...
           command: |
             apt-get -qq update
-            apt-get -qy install python-sphinx python-pip
+            apt-get -qy install python-sphinx python-pip codespell
       - run: *setup_prerelease_commit_hash
       - run:
           name: Build documentation
           command: ./scripts/docs.sh
+      - run:
+          name: Check spelling
```
Comment (axic) [inline .circleci/config.yml]: I don't think we should run this twice (on linux and mac). Actually it would be great if it could be moved to its own job (like the `docs` job) and as such would show up separately on github.

File: libsolidity/inlineasm/AsmParser.cpp [hunk 3]
```diff
@@ -150,7 +150,7 @@ assembly::Statement Parser::parseStatement()
 			expectToken(Token::Comma);
 			elementary = parseElementaryOperation();
 			if (elementary.type() != typeid(assembly::Identifier))
-				fatalParserError("Variable name expected in multiple assignemnt.");
+				fatalParserError("Variable name expected in multiple assignment.");
```
Comment (axic) [inline libsolidity/inlineasm/AsmParser.cpp]: Can you please pull this fix (together with its test) into a separate PR?

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -1848,7 +1848,7 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		if (functionType->takesArbitraryParameters())
 			m_errorReporter.typeError(
 				_functionCall.location(),
-				"Named arguments cannnot be used for functions that take arbitrary parameters."
+				"Named arguments cannot be used for functions that take arbitrary parameters."
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: Is there no test case for this? Also please pull this into a separate PR (can be shared with the other I proposed just above).

File: libsolidity/inlineasm/AsmParser.cpp (hunk 3 above)
Comment (cryptomental) [reply libsolidity/inlineasm/AsmParser.cpp]: Ok, will be pulled out.

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (cryptomental) [reply libsolidity/analysis/TypeChecker.cpp]: Will be pulled out together with the one above.

File: .circleci/config.yml (hunk 2 above)
Comment (cryptomental) [reply .circleci/config.yml]: All right, will be moved to a separate job.

File: scripts/codespell_whitelist.txt (hunk 1 above)
Comment (cryptomental) [reply scripts/codespell_whitelist.txt:3]: Yes. libsolidity/codegen/ABIFunctions.cpp and test/libsolidity/SolidityEndToEndTest.cpp .
The former is a template tag, the latter a specific test that has to contain "nd".

File: libsolidity/inlineasm/AsmParser.cpp (hunk 3 above)
Comment (cryptomental) [reply libsolidity/inlineasm/AsmParser.cpp]: Pulled out. https://github.com/ethereum/solidity/pull/4464 opened.

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (cryptomental) [reply libsolidity/analysis/TypeChecker.cpp]: Pulled out. https://github.com/ethereum/solidity/pull/4464 opened.

File: .circleci/config.yml (hunk 2 above)
Comment (cryptomental) [reply .circleci/config.yml]: @axic done. I moved it to a separate job. Currently it is failing and it is supposed to fail, please see https://circleci.com/gh/ethereum/solidity/22005?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link as I pulled out the two changes into https://github.com/ethereum/solidity/pull/4464 , which should be merged first.

**Comment by cryptomental** [comment]
@axic only the test_check_passing is failing, it will not fail anymore after https://github.com/ethereum/solidity/pull/4464 is merged. I hope everything is ready to be merged now. Please continue the review.

File: .circleci/config.yml [hunk 5]
```diff
@@ -158,6 +158,25 @@ jobs:
...
+      TERM: xterm
+    steps:
+      - checkout
+      - attach_workspace:
+          at: build
+      - run:
+          name: Install dependencies
+          command: |
+            apt-get -qq update
+            apt-get -qy install python-pip
```
Comment (chriseth) [inline .circleci/config.yml]: Does it makes sense to use `- image: circleci/python:[TAG]` here which might already include `pip`?

**Comment by chriseth** [comment]
Please rebase.

**Comment by cryptomental** [comment]
@chriseth I rebased and had to correct docs/solidity-by-example.rst as it went in with the latest PR.
Also circleci/python:3.6 is used as a base image for spell checker.

File: .circleci/config.yml (hunk 2 above)
Comment (chriseth) [reply .circleci/config.yml]: Did you mean to remove this? Does it something else than the job above?

Comment (cryptomental) [reply .circleci/config.yml]: @chriseth Removed as well. Only a separate job remains now.

File: docs/style-guide.rst [hunk 6]
```diff
@@ -188,8 +188,8 @@ Event Definitions and Event Emitters
 Yes::
 
     event LongAndLotsOfArgs(
-        adress sender,
-        adress recipient,
+        address sender,
```
Comment (chriseth) [inline docs/style-guide.rst:191]: This is gold!

File: libsolidity/codegen/CompilerUtils.cpp [hunk 7]
```diff
@@ -364,7 +364,7 @@ void CompilerUtils::encodeToMemory(
 	// Stack during operation:
 	// <v1> <v2> ... <vn> <mem_start> <dyn_head_1> ... <dyn_head_r> <end_of_mem>
 	// The values dyn_head_i are added during the first loop and they point to the head part
-	// of the ith dynamic parameter, which is filled once the dynamic parts are processed.
+	// of the with dynamic parameter, which is filled once the dynamic parts are processed.
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.cpp]: Sorry, but this was correct. Perhaps replace by `nth`?

Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp]: But then also replace `dyn_head_i` by `dyn_head_n`.

File: scripts/codespell_whitelist.txt (hunk 1 above)
Comment (chriseth) [reply scripts/codespell_whitelist.txt:3]: So it's not case sensitive? Perhaps you can change the test so that it does not contain nd? Where exactly does it occur?

Comment (cryptomental) [reply scripts/codespell_whitelist.txt:3]: It occurs here: https://github.com/ethereum/solidity/blob/develop/test/libsolidity/SolidityEndToEndTest.cpp#L7779
The test case mentions "a specific bug" and I decided not to touch it not knowing what the bug was.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 7 above)
Comment (cryptomental) [reply libsolidity/codegen/CompilerUtils.cpp]: @chriseth done.

**Comment by cryptomental** [comment]
AppVeyor still did not kick in. I am not sure if this is normal to wait for it so many hours.

File: docs/style-guide.rst (hunk 6 above)
Comment (cryptomental) [reply docs/style-guide.rst:191]: Especially that this was found in the style guide.

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
I think we should have put `tests/compilationTests` on an exclusion list.

@cryptomental is that possible at all?
